File: week4/ch.Choe/Week4/Week4/Pizza/Pizza.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Week4.Ingredient;
using Week4.IngredientFactory;

namespace Week4
{
    public abstract class Pizza
    {
        protected string m_Name;
        protected Dough m_Dough;
        protected Cheese m_Cheese;
        protected Pepperoni m_Pepperoni;
        protected PizzaIngredientFactory m_Factory;
        protected IPizzaCutting m_Cutting;
        public Pizza(string name, PizzaIngredientFactory fac, IPizzaCutting cut)
        {
            m_Name = name;
            m_Factory = fac;
            m_Cutting = cut;
        }

        public abstract void Prepare();
        
        public void Bake()
        {
            Console.WriteLine("Baking... ");
        }

        public void Cut()
        {
            m_Cutting.Cutting();
        }
        public virtual void Box()
        {
            Console.WriteLine("Box");
        }
        public string GetName()
        {
            return m_Name;
        }

    }
}

File: week4/minj0i/Pizza.java
package week4.minj0i;

import java.util.ArrayList;

public abstract class Pizza {
	String name;
	String dough;
	String sauce;
	ArrayList toppings = new ArrayList<>();
	
	void prepare() {
		System.out.println("준비중 " + name);
		System.out.println("Tossing 도우.");
		System.out.println("소스 추가......");
		System.out.println("토핑 추가: ");
		for (int i=0; i<toppings.size(); i++) {
			System.out.println(" " + toppings.get(i));
		}
	}
	
	void bake() {
		System.out.println("350도에서 25분동안 굽는다");
	}
	
	void cut() {
		System.out.println("피자를 자른다");
	}
	
	void box() {
		System.out.println("우리의 피자박스에 피자를 넣는다");
	}
	
	public String getName() {
		return name;
	}
}
